design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+```ts
+class ApplicationLoadBalancer {
+  serialize(ctx: ISerializationContext): void {
+    ctx.writeString('LoadBalancerArn', this.loadBalancerArn);
+    ctx.writeObject('SecurityGroup', this.securityGroup);
+  }
+
+  static deserializeApplicationLoadBalancer(ctx: IDeserializationContext): IApplicationLoadBalancer {
+    const exists = ctx.scope.findChild(ctx.id);
+    if (exists) {
+      return exists;
+    }
+
+    return new ImportedApplicationLoadBalancer(ctx.scope, ctx.id, {
+      loadBalancerArn: ctx.readString('LoadBalancerArn'),
+      securityGroup: ec2.SecurityGroup.deserializeSecurityGroup(ctx.readObject('SecurityGroup'))
+    });
+  }
+}
+```
+
+### Imports/Exports
+
+Now that constructs can be serialized and deserialized into a key-value context, we can implement a serialization mechanism for exporting resources from stacks and importing them in another stack. Importing can be done either at synthesis time using an environmental context provider or at runtime by returning `Fn::ImportValue` tokens for `readString`.
+
+The following methods will be added to the `Stack` class:
+
+```ts
+class Stack {
+  exportString(exportName: string, value: string, options?: ExportOptions): void;
+  importString(exportName: string, options?: ImportOptions): string;
+}
+
+interface ExportOptions {
+  description?: string
+}
+
+interface ImportOptions {
+  type?: ResolveType // default is Synthesis
+  weak?: boolean; // default is "false"
+}
+
+enum ResolveType {
+  Synthesis,
+  Deployment
+}
+```
+
+The `exportString` method creates an AWS CloudFormation Output for this value assigned to this export name.
+
+The `importString` method returns the value for this specific export name. When importing a value, users can specify the following `ImportOptions`:
+
+* If `type` is set to `Deployment`, the method will return an `Fn::ImportValue(exportName)` token. This means that CDK code cannot reason about the concrete value, which will only be resolved when the stack is deployed.
+* If `type` is set to `Synthesis` (default) the method will exercise an environmental context provider to look up the export value __during synthesis__. The concrete value will be propagated to the CDK app and can be reasoned about like any normal value.
+* The `weak` option is only relevant for synth-time resolution. If it is `false` (which is the default), the CDK will automatically embed a `Metadata` entry on the consuming resource with an `Fn::ImportValue`. This will force CloudFormation to take a strong reference on the export, even through the actual value is concretely resolved during synthesis. This ensures, for example, that the producing stack can't be deleted as long as there stacks consuming the exported values. This behavior (which is the default), can be disabled by settings `weak: true`, in which case the `Fn::ImportValue` will simply not be included.
```

Comment:
Feels like maybe we get a more coherent overall surface if you remove `weak` and just have `ResolveType` be one of `Synthesis`, `WeakSynthesis`, `Deployment`.